### PR TITLE
feat: cancel queries on navigation-ish

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

We have query cancellation that fires when someone changes filters while a slow query is running.

But, they may become frustrated and navigate away within the app. We would leave any ClickHouse queries running in that case.

## Changes

Calls into the mounted dashboard or insight logic when the `Dashboard.tsx` or `Insight.tsx` is unmounted and attempts to cancel queries

NB: this doesn't try to unify this cancellation mechanism between insights and dashboards, just adds the new reasons to call it

## How did you test this code?

running locally and 👀

### CODE REVIEW

File: frontend/src/scenes/insights/Insight.tsx [hunk 1]
```diff
@@ -72,6 +73,9 @@ export function Insight({ insightId }: { insightId: InsightShortId | 'new' }): J
 
     useEffect(() => {
```
Comment (macobo) [inline frontend/src/scenes/insights/Insight.tsx:74]: Q: Does this component get unmounted when moving from trends -> funnels?

File: frontend/src/scenes/insights/Insight.tsx [hunk 2]
```diff
@@ -72,6 +73,9 @@ export function Insight({ insightId }: { insightId: InsightShortId | 'new' }): J
...
         reportInsightViewedForRecentInsights()
+        return () => {
+            abortAnyRunningQuery()
+        }
     }, [insightId])
```
Comment (macobo) [inline frontend/src/scenes/insights/Insight.tsx:79]: What effect does this useEffect depending on insightId have?

Should we make it agnostic?

**Comment by macobo** [comment]
Would be great to confirm in what navigation scenarios queries get cancelled now and in which they don't.

File: frontend/src/scenes/dashboard/Dashboard.tsx [hunk 3]
```diff
@@ -51,10 +51,14 @@ function DashboardScene(): JSX.Element {
         dashboardMode,
         receivedErrorsFromAPI,
     } = useValues(dashboardLogic)
-    const { setDashboardMode, setDates, reportDashboardViewed, setProperties } = useActions(dashboardLogic)
+    const { setDashboardMode, setDates, reportDashboardViewed, setProperties, abortAnyRunningQuery } =
+        useActions(dashboardLogic)
 
     useEffect(() => {
         reportDashboardViewed()
+        return () => {
```
Comment (macobo) [inline frontend/src/scenes/dashboard/Dashboard.tsx:59]: Suggestion: create a `useQueryCancellation(fn)`

Comment (pauldambra) [reply frontend/src/scenes/dashboard/Dashboard.tsx:59]: The problem with wrapping (or at least the problem I've not figured out the solution to) is the insight and dashboard cancellation cases are still really different

**Comment by pauldambra** [comment]
> Would be great to confirm in what navigation scenarios queries get cancelled now and in which they don't.

So, we have the existing behaviour that catches changes in filters and cancels running queries.

This adds cancelling queries when the components are no longer in the DOM.

So the cases covered right now

* when an insight is running and you navigate to another page
* when an insight is running and you change anything that causes a new query to start
* when a dashboard is running and you navigate to another page
* when a dashboard is running and you change the time range or a filter

File: frontend/src/scenes/insights/Insight.tsx [hunk 4]
```diff
@@ -74,6 +82,19 @@ export function Insight({ insightId }: { insightId: InsightShortId | 'new' }): J
         reportInsightViewedForRecentInsights()
     }, [insightId])
 
+    useEffect(() => {
+        if (showErrorMessage) {
```
Comment (macobo) [inline frontend/src/scenes/insights/Insight.tsx:86]: Is this an unrelated change? Context?

File: frontend/src/scenes/insights/insightLogic.ts [hunk 5]
```diff
@@ -295,11 +296,14 @@ export const insightLogic = kea<insightLogicType>([
                     // fetch this now, as it might be different when we report below
                     const scene = sceneLogic.isMounted() ? sceneLogic.values.scene : null
 
-                    // If a query is in progress, kill that query
                     if (cache.abortController) {
```
Comment (macobo) [inline frontend/src/scenes/insights/insightLogic.ts:299]: Could we dispatch abortAnyRunningQuery here? Or extract function?

**Review by macobo — APPROVED** [review]

File: frontend/src/scenes/dashboard/dashboardLogic.tsx [hunk 6]
```diff
@@ -995,10 +987,7 @@ export const dashboardLogic = kea<dashboardLogicType>([
                 }
 
                 refreshesFinished += 1
-                if (refreshesFinished === insights.length) {
-                    breakpoint()
-                    cache.abortController = null
-
+                if (!cancelled && refreshesFinished === insights.length) {
```
Comment (pauldambra) [inline frontend/src/scenes/dashboard/dashboardLogic.tsx:990]: @macobo one last thing to check now I have the tests passing

I've had to add a `!cancelled` gate here because sometimes this conditional was executed after tests had finished running.

I think it's OK not to report time_to_see data for the dashboard load here since we'd be reporting on the cancellation but....

Comment (macobo) [reply frontend/src/scenes/dashboard/dashboardLogic.tsx:990]: Thanks for the heads up.
